cpp/include/raft/common/nvtx.hpp

```diff
@@ -0,0 +1,125 @@
+{
+  detail::pushRange(format, args...);
+}
+
+/**
+ * @brief Synchronize CUDA stream and push a named nvtx range
+ * @param format range name format (accepts printf-style arguments)
+ * @param args the arguments for the printf-style formatting
+ * @param stream stream to synchronize
+ */
+template <typename... Args>
+inline void PUSH_NVTX_RANGE(rmm::cuda_stream_view stream, const char* format, Args... args)
+{
+  detail::pushRange(stream, format, args...);
+}
+
+/** Pop the latest range */
+inline void POP_NVTX_RANGE() { detail::popRange(); }
+
+/**
+ * @brief Synchronize CUDA stream and pop the latest nvtx range
+ * @param stream stream to synchronize
+ */
+inline void POP_NVTX_RANGE(rmm::cuda_stream_view stream) { detail::popRange(stream); }
+
+/** Push a named nvtx range that would be popped at the end of the object lifetime. */
+class NvtxRange {
+ private:
+  std::optional<rmm::cuda_stream_view> streamMaybe;
+
+  /* This object is not meant to be touched. */
+  NvtxRange(const NvtxRange&) = delete;
+  NvtxRange(NvtxRange&&)      = delete;
+  NvtxRange& operator=(const NvtxRange&) = delete;
+  NvtxRange& operator=(NvtxRange&&)        = delete;
+  static void* operator new(std::size_t)   = delete;
+  static void* operator new[](std::size_t) = delete;
+
+ public:
+  /**
+   * Synchronize CUDA stream and push a named nvtx range
+   * At the end of the object lifetime, synchronize again and pop the range.
+   *
+   * @param stream stream to synchronize
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  NvtxRange(rmm::cuda_stream_view stream, const char* format, Args... args)
+    : streamMaybe(std::make_optional(stream))
+  {
+    PUSH_NVTX_RANGE(stream, format, args...);
+  }
+
+  /**
+   * Push a named nvtx range.
+   * At the end of the object lifetime, pop the range back.
+   *
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  NvtxRange(const char* format, Args... args) : streamMaybe(std::nullopt)
+  {
+    PUSH_NVTX_RANGE(format, args...);
+  }
+
+  ~NvtxRange()
+  {
+    if (streamMaybe.has_value()) {
+      POP_NVTX_RANGE(*streamMaybe);
+    } else {
+      POP_NVTX_RANGE();
+    }
+  }
+};
+
+/*!
+  \def RAFT_USING_NVTX_RANGE(...)
+  When NVTX is enabled, push a named nvtx range and pop it at the end of the enclosing code block.
+
+  This macro initializes a dummy NvtxRange variable on the stack,
+  which pushes the range in its constructor and pops it in the destructor.
+*/
+#ifdef NVTX_ENABLED
+#define _RAFT_USING_NVTX_RANGE_CAT(a, b)       a##b
+#define _RAFT_USING_NVTX_RANGE_UNIQUE_NAME(ln) _RAFT_USING_NVTX_RANGE_CAT(_NvtxRange_, ln)
+#define RAFT_USING_NVTX_RANGE(...) \
+  raft::common::NvtxRange _RAFT_USING_NVTX_RANGE_UNIQUE_NAME(__LINE__)(__VA_ARGS__)
```

Comment:
The use of the line information and variable args here is a compelling reason enough for me to be in favor of using a macro here. This is also the reason we use macros for logging. I could go either way with the `NVTX_ENABLED` variable but I'm inclined to use this pattern since it's familiar and used throughout the RAPIDS c/c++ stack. @wphicks does exposing the macro as a wrapper just to capture line / args info seem like a reasonable compromise here? 

I do like the use of the macros as a direct wrapper around a C++ function, which should make it easier to maintain.

Reply:
Note, I use the line info here only to give a unique name to the variable, it's not passed along to the constructor. Also the arguments are passed as-is, so it's really only about the convenience of using the macro vs the clarity of declaring an `NvtxRange` variable of explicitly.

Reply:
(Just to clarify, I use these three macros here instead of one to make sure `__LINE__` is expanded before it's concatenated with `##`)

Reply:
@cjnolet I think given that the line info isn't really exposed in the reported NVTX info, I'd still prefer to avoid it (unless someone sees some other benefit that I'm not?).

Reply:
Also please feel free to just overrule me if there's consensus in favor of the macro. I'm trying to reduce our macro usage wherever possible, but this is a _great_ contribution with or without. So, let's get it in one way or another! :smile:

Reply:
If the line info is never shown to the user, I do agree with Will here that I don’t see a point to using a macro over just invoking the function directly.

Reply:
With that said, in practice, I'm also somewhat impartial here, so @achirkin, if you prefer to use the macro, I'm okay with keeping it in here.